### PR TITLE
Add orderly workflows and yml reading

### PR DESCRIPTION
**Pull request by r-ash**
This PR will

* Add skeleton `orderly_workflow` function
* Adds yml reading

Doesn't actually do any report running but locates a yml file, reads it and parses this into a list of `steps`. 

Going to add some validation then this is ready to review.

### CODE REVIEW

File: R/workflow.R [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    }
+  )
+)
+
+parse_steps <- function(yml) {
+  unname(unlist(yml$steps))
+}
+
+validate_steps <- function(steps, config, workflow_name) {
+  reports <- basename(list_dirs(path_src(config$root)))
```
Comment (r-ash) [inline R/workflow.R:68]: This is the sample impl as `orderly_list` but using the config object. Looks like the impl of the `orderly_list` stuff if half way through a move to a `list2` function so wanted to mention this. Don't know whether it is worth adding an `orderly_list(config)` function which is called both here and in `orderly_list(root, locate)`? Or some other way of tidying up the code re-use

Comment (richfitz) [reply R/workflow.R:68]: yeah, there's a bunch of stuff like this that needs tidying up. I'm happy for you to tidy up here or in another ticket. It's internal stuff only so we can do it however we want

Comment (richfitz) [reply R/workflow.R:68]: Or you can just use `orderly_list(config)` which works fine as `orderly_config_get(config)` just returns `config`

File: R/workflow.R [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    initialize = function(workflow_name, workflow_path, config) {
+      self$workflow_name <- workflow_name
+      self$workflow_id <- ids::random_id()
+      self$config <- config
+      raw <- yaml_read(workflow_path)
+      self$steps <- parse_steps(raw)
+      validate_steps(self$steps, self$config, workflow_name)
+    },
+
+    run = function() {
```
Comment (richfitz) [inline R/workflow.R:55]: Consider making this just error while we don't have an implementation?

File: R/workflow.R [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+      ids::random_id()
+    }
+  )
+)
+
+parse_steps <- function(yml) {
+  unname(unlist(yml$steps))
+}
+
+validate_steps <- function(steps, config, workflow_name) {
```
Comment (richfitz) [inline R/workflow.R:67]: I think this should use `check_fields` to validate that each step has just the configuration options we support

File: R/workflow.R [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  reports <- basename(list_dirs(path_src(config$root)))
+  for (step in steps) {
```
Comment (richfitz) [inline R/workflow.R:69]: perhaps report all back at once?  Elsewhere we use

```
msg <- setdiff(step, reports)
if (length(msg)) {
  stop(sprintf("Cannot run workflow '%s' as report missing: %s",
               paste(squote(msg), collapse = ", ")))
}
```

File: R/workflow.R (hunk 1 above)
Comment (r-ash) [reply R/workflow.R:68]: Oh nice - I missed that. `orderly_list(config)` is perfect.

File: R/workflow.R (hunk 3 above)
Comment (r-ash) [reply R/workflow.R]: Nice, added that now too.

**Pull request merged**